src/config.rs

```diff
@@ -0,0 +1,385 @@
+        // variable can be different types. This way we can make type inference
+        // figure out what try_into() deserializes to.
+        macro_rules! get_and_insert {
+            ($table:expr, $key:expr => $out:expr) => {
+                if let Some(value) = $table.remove($key).and_then(|v| v.try_into().ok()) {
+                    $out = value;
+                }
+            };
+        }
+
+        get_and_insert!(table, "title" => cfg.book.title);
+        get_and_insert!(table, "authors" => cfg.book.authors);
+        get_and_insert!(table, "source" => cfg.book.src);
+        get_and_insert!(table, "description" => cfg.book.description);
+
+        // This complicated chain of and_then's is so we can move 
+        // "output.html.destination" to "book.build_dir" and parse it into a 
+        // PathBuf.
+        let destination: Option<PathBuf> = table.get_mut("output")
+            .and_then(|output| output.as_table_mut())
+            .and_then(|output| output.get_mut("html"))
+            .and_then(|html| html.as_table_mut())
+            .and_then(|html| html.remove("destination"))
+            .and_then(|dest| dest.try_into().ok());
+
+        if let Some(dest) = destination {
+            cfg.book.build_dir = dest;
+        }
+
+        cfg.rest = table;
+        cfg
+    }
+}
+
+fn recursive_get<'a>(key: &[&str], table: &'a Table) -> Option<&'a Value> {
+    if key.is_empty() {
+        return None;
+    } else if key.len() == 1 {
+        return table.get(key[0]);
+    }
+
+    let first = key[0];
+    let rest = &key[1..];
+
+    if let Some(&Value::Table(ref nested)) = table.get(first) {
+        recursive_get(rest, nested)
+    } else {
+        None
+    }
+}
+
+fn recursive_get_mut<'a>(key: &[&str], table: &'a mut Table) -> Option<&'a mut Value> {
+    // TODO: Figure out how to abstract over mutability to reduce copy-pasta
+    if key.is_empty() {
+        return None;
+    } else if key.len() == 1 {
+        return table.get_mut(key[0]);
+    }
+
+    let first = key[0];
+    let rest = &key[1..];
+
+    if let Some(&mut Value::Table(ref mut nested)) = table.get_mut(first) {
+        recursive_get_mut(rest, nested)
+    } else {
+        None
+    }
+}
+
+impl<'de> Deserialize<'de> for Config {
+    fn deserialize<D: Deserializer<'de>>(de: D) -> ::std::result::Result<Self, D::Error> {
+        let raw = Value::deserialize(de)?;
+
+        let mut table = match raw {
+            Value::Table(t) => t,
+            _ => {
+                use serde::de::Error;
+                return Err(D::Error::custom(
+                    "A config file should always be a toml table",
+                ));
+            }
+        };
+
+        if is_legacy_format(&table) {
+            warn!("It looks like you are using the legacy book.toml format.");
+            warn!("We'll parse it for now, but you should probably convert to the new format.");
+            warn!("See the mdbook documentation for more details");
```

Comment:
The explicit differences are mentioned neither here nor in the docs. I think it would be a nice gesture to just say:

> In short words: Just move all top level configuration entries like `title`, `author` and `description` into a group called `[book]`.

Reply:
Good point. I've seen `rustc` and other tools emit messages along the same line of "this is going to stop working soon so update your stuff. If you do x, y, and z then that should fix 99% of your problems" and found it a really nice gesture.